fdbserver/DataDistribution.actor.cpp

```diff
@@ -2490,6 +2532,83 @@ ACTOR Future<Void> removeBadTeams(DDTeamCollection* self) {
 	return Void();
 }
 
+bool inCorrectDC(DDTeamCollection* self, TCServerInfo* server) {
+	return (self->includedDCs.empty() ||
+	        std::find(self->includedDCs.begin(), self->includedDCs.end(), server->lastKnownInterface.locality.dcId()) !=
+	            self->includedDCs.end());
+}
+
+// Is there any healthy team whose members do not include serverID
+bool existOtherHealthyTeams(DDTeamCollection* self, UID serverID) {
+	for (auto& team : self->teams) {
+		if (team->isHealthy() && std::count(team->serverIDs.begin(), team->serverIDs.end(), serverID) == 0) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+ACTOR Future<Void> removeWrongStoreType(DDTeamCollection* self) {
+	// Wait for storage servers to initialize its storeType
+	wait(delay(SERVER_KNOBS->DD_REMOVE_STORE_ENGINE_DELAY));
+
+	state bool foundSSToRemove = false;
+	state Reference<TCServerInfo> secondPreferedSSToRemove;
+
+	loop {
+		foundSSToRemove = false;
+		secondPreferedSSToRemove = Reference<TCServerInfo>();
+		if (self->doRemoveWrongStoreType.get() == false) {
+			// Once the wrong storeType SS picked to be removed is removed, doRemoveWrongStoreType will be set to true;
+			// In case the SS fails in between, we should time out and check for the next SS.
+			wait(self->doRemoveWrongStoreType.onChange() || delay(SERVER_KNOBS->DD_REMOVE_STORE_ENGINE_TIMEOUT));
```

Comment:
do not rely on the timeout, and change to while(self->doRemoveWrongStoreType.get() == false) to ensure you do not continue if it is false